### PR TITLE
Major revamp of the code structures

### PR DESCRIPTION
**Pull request by tousifislam**
1. moved all general scripts to common utils
2. deleted directories within model_utils
3. model utils now have generic scripts for both splines and GPR
4.  made nr calibration generic for all models
5.  added checks on user inputs, bounds etc
6. cleaned up all codes

### CODE REVIEW

File: surrogates/common_utils/doc_string.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    ## -------------------------------------------------------------------------- ##
+    ## general overview
+    ## -------------------------------------------------------------------------- ##
+    
+    Description : warpper to generate BHPT surrogate waveforms
+    
+    Input
+    =====
+    q:      mass ratio
+
```
Comment (sfield17) [inline surrogates/common_utils/doc_string.py:31]: maybe specify "mass ratio (with q >=1)"

File: surrogates/common_utils/doc_string.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    
+    Description : warpper to generate BHPT surrogate waveforms
+    
+    Input
+    =====
+    q:      mass ratio
+    
+    spin1: spin of the primary black hole / Default: None
+           Not Implemented in most models
+
```
Comment (sfield17) [inline surrogates/common_utils/doc_string.py:35]: specify dimensionless spin where  -1 <= spin1 <= 1.

possibly even change the name to chi1, since spin1 is likely to be interpreted as dimensioned.

File: surrogates/common_utils/doc_string.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    Input
+    =====
+    q:      mass ratio
+    
+    spin1: spin of the primary black hole / Default: None
+           Not Implemented in most models
+    
+    spin2: spin of the secondary black hole / Default: None
+           Not Implemented in any model so far
+
```
Comment (sfield17) [inline surrogates/common_utils/doc_string.py:39]: same comment as for spin1

File: surrogates/common_utils/doc_string.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+                  (7,7),(8,6),(8,7),(8,8),(9,7),(9,8),(9,9),(10,8),(10,9)]
+             for BHPTNRSur1dq1e4
+            
+    M_tot:     total mass of the binary in solar masses
+               Default: None (in which case a geometric waveform is returned)
+    
+    dist_mpc:  distance of the binary from the observer in Mpc
+               Default: None (in which case geometric wf is returned)
+               
+    orb_phase: orbital phase
```
Comment (sfield17) [inline surrogates/common_utils/doc_string.py]: we need to specify what the convention is. same for inclination angle

File: surrogates/common_utils/doc_string.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    
+    inclination: inclination angle
+    
+    lmax:  5 (default)
+           Modes upto l=5 are NR calibrated. Modes l>6 are uncalibrated
+           Note: If one provides a list of modes, modes beyond lmax 
+                 will not be returned
```
Comment (sfield17) [inline surrogates/common_utils/doc_string.py]: specify that both positive and negative m modes are computed (gwsurrogate doesn't do this, for example, so there's a different behavior!)

File: surrogates/common_utils/doc_string.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    
+    dist_mpc:  distance of the binary from the observer in Mpc
+               Default: None (in which case geometric wf is returned)
+               
+    orb_phase: orbital phase
+    
+    inclination: inclination angle
+    
+    lmax:  5 (default)
+           Modes upto l=5 are NR calibrated. Modes l>6 are uncalibrated
```
Comment (sfield17) [inline surrogates/common_utils/doc_string.py]: l=5 is a choice for the current model, right? Might be better to remind the user that (i) modes are only calibrated to NR up to some maximum value of ell, and (ii) consult a specific model's documentation for information on how modes are calibrated to NR.

File: surrogates/common_utils/doc_string.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+            
+    mode_sum:  If true, all modes are summed. If false all modes are returned 
+               in a dictionary. Default: false
+               Note: Only works when orb_phase and inclination are not None
+               
+    calibrated:  Whether you want NR-calibrated waveform or not
+                 When set to True, it applies a scaling to the uncalibrated
+                 surrogate waveform. This scaling has been obtained by calibrating
+                 the ppBHPT waveforms to NR in comparable mass ratio
+                 regime (3<=q<=9). If set to False, the raw (uncalibrated)
```
Comment (sfield17) [inline surrogates/common_utils/doc_string.py]: 3<=q<=9 is a choice for the current model, right? Might be better to remind the user that (i) calibration to NR is done over a region of parameter space and (ii) consult a specific model's documentation for information on how modes are calibrated to NR.

File: surrogates/common_utils/fits.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+import numpy as np
+import scipy
+from scipy.interpolate import splrep, splev
+from . import utils
+from .eval_pysur import evaluate_fit as evaluate_GPR
+
+#----------------------------------------------------------------------------------------------------
+def evaluate_GPR_at_EIM_nodes(X, fit_data):
+    """ Evaluate the spline at one EIM node """
+    return None
```
Comment (sfield17) [inline surrogates/common_utils/fits.py]: use "raise NotImplementedError" instead of "return None"

File: surrogates/common_utils/fits.py [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+def evaluate_splines_at_EIM_nodes(X, fit_data):
+    """ Evaluate the spline at one EIM node """
+    
+    [h_eim_spline, eim_indicies] = fit_data
+    return np.array([splev(X, h_eim_spline[j]) for j in range(len(eim_indicies))])
+
+
+#----------------------------------------------------------------------------------------------------
+def EIM_B_to__waveform_datapiece(B, eim_vals):
+    """ Compute the interpolated waveform for a single mode : dominant 22 mode """
```
Comment (sfield17) [inline surrogates/common_utils/fits.py]: does this routine only work for the 22 mode? seems more generic than that -- looks like any data piece (amp, phase, etc). maybe reword the documentation to mention its for generic data pieces (e.g. the 22 mode's amplitude and phase)

File: surrogates/common_utils/fits.py [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+    # combine h_eim and  eim basis matrix to give full datapiece
+    h_approx_datapiece = EIM_B_to__waveform_datapiece(B, h_eim_datapiece) 
+    
+    return h_approx_datapiece
+
+
+#----------------------------------------------------------------------------------------------------
+def evaluate_surrogate_mode(X, fit_data_1, fit_data_2, B_datapiece_1, B_datapiece_2, 
+                            fit_func, decomposition_func, norm):
+    """ Compute the interpolated waveform for a single mode : higher order modes"""
```
Comment (sfield17) [inline surrogates/common_utils/fits.py]: I think more documentation is needed to explain what this function is doing. In particular, I am confused about...

1. does this function not work for the 22 mode? If not, why not? What routine is used to compute the 22 mode?
2. this looks like an "internal" function that isn't supposed to be called outside of all_modes_surrogate. As such, I would suggest adding an underscore to the function, so _evaluate_surrogate_mode. This will make it clear that its used for internal stuff (ie its a "private" function). This is used especially for private methods of classes, but I think its helpful here too (https://peps.python.org/pep-0008/#method-names-and-instance-variables). Other functions in this python file that shouldn't be called directly by the user could also get an underscore.
3. I would mention in the documentation that the user should look at all_modes_surrogate for information on the inputs.

File: surrogates/common_utils/fits.py [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+    return h_approx
+
+
+#----------------------------------------------------------------------------------------------------
+def all_modes_surrogate(modes, X_input, fit_data_dict_1, fit_data_dict_2, \
+                        B_dict_1, B_dict_2, lmax, fit_func, decomposition_funcs, norm):
+
+    """ Takes the fit data (either from splines or GPR), matrix B and computes the 
+        interpolated waveform for all modes 
+
```
Comment (sfield17) [inline surrogates/common_utils/fits.py:90]: I would add. "this routine computes uncalibrated waveforms in geometric units, and is meant to be called from evaluate_surrogtes.evaluate_surrogate"

File: surrogates/common_utils/fits.py [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+            fit_data_2 = fit_data_dict_2[mode]
+            
+            # read the decompositon function for the modes; special treatment for the
+            # 22 mode and higher order modes
+            if mode==(2,2):
+                decomposition_func = decomposition_funcs[0]
+            else:
+                decomposition_func = decomposition_funcs[1]
+    
+            # evaluate surrogate modes
```
Comment (sfield17) [inline surrogates/common_utils/fits.py:151]: Is my understanding correct that for models using the co-orbital frame, the modes are still in the co-oorbital frame at this point? And the rotation back to the inertial frame is in evaluate_surrogate? If so, adding that code comment here would be helpful.

File: surrogates/common_utils/utils.py [hunk 13]
```diff
@@ -1,38 +1,73 @@
...
+            orbital_phase = np.unwrap(np.angle(h_coorb[mode]))/2
+            h_inertial[mode] = h_coorb[mode]
+        else:
+            # transform HMs to inertial frame
+            h_inertial[mode] = h_coorb[mode]*np.exp(1j*m*np.array(orbital_phase))
+            
+    return h_inertial
 
 #---------------------------------------------------------------------------------------------------- 
 def geo_to_SI(t_geo, h_geo, M_tot, dist_mpc):
```
Comment (sfield17) [inline surrogates/common_utils/utils.py]: maybe this routine should be moved to gwtools? Its now used both here and gwsurrogate. Its also generally useful so a stand-alone function could make sense

File: surrogates/common_utils/utils.py [hunk 14]
```diff
@@ -1,38 +1,73 @@
...
     strain_geo_to_SI = (G*M/C_SI**3)/dL
     h_SI={}
     for mode in h_geo.keys():
         h_SI[(mode)] = np.array(h_geo[mode])*strain_geo_to_SI
     
     return t_SI, h_SI
 
-
 #---------------------------------------------------------------------------------------------------- 
 def phase_rotation(h, delta_orb_phase):
```
Comment (sfield17) [inline surrogates/common_utils/utils.py:48]: maybe this routine should be moved to gwtools? Its now used both here and gwsurrogate. Its also generally useful so a stand-alone function could make sense

File: surrogates/model_utils/load_surrogates.py [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+             (5,3),(5,4),(5,5),(6,4),(6,5),(6,6),(7,5),(7,6),(7,7),
+             (8,6),(8,7),(8,8),(9,7),(9,8),(9,9),(10,8),(10,9)]
+    
+    # modes used in nr calibration
+    nrcalib_modes = [(2,2),(3,3),(4,4),(5,5)]
+    
+    # obtain all fit data
+    time, fit_data_dict_1, fit_data_dict_2, B_dict_1, B_dict_2, alpha_coeffs, beta_coeffs \
+                    = load_spl.load_surrogate(h5_data_dir, fname, wf_modes, nrcalib_modes)
+
```
Comment (sfield17) [inline surrogates/model_utils/load_surrogates.py:77]: should this be integrated more closely with gwsurrogate? Or is it not worth the effort?

**Review by sfield17 — CHANGES_REQUESTED** [review]

File: surrogates/common_utils/doc_string.py (hunk 1 above)
Comment (tousifislam) [reply surrogates/common_utils/doc_string.py:31]: Done!

File: surrogates/common_utils/doc_string.py (hunk 2 above)
Comment (tousifislam) [reply surrogates/common_utils/doc_string.py:35]: Done!

File: surrogates/common_utils/doc_string.py (hunk 3 above)
Comment (tousifislam) [reply surrogates/common_utils/doc_string.py:39]: Done!

File: surrogates/common_utils/doc_string.py (hunk 4 above)
Comment (tousifislam) [reply surrogates/common_utils/doc_string.py]: Done. Did an experiment to figure out whether it is the phase at the start or at the merger. 
Result : it is the phase at the start of the waveform.

File: surrogates/common_utils/doc_string.py (hunk 5 above)
Comment (tousifislam) [reply surrogates/common_utils/doc_string.py]: Done!

File: surrogates/common_utils/doc_string.py (hunk 6 above)
Comment (tousifislam) [reply surrogates/common_utils/doc_string.py]: Good point. Made the change.

File: surrogates/common_utils/doc_string.py (hunk 7 above)
Comment (tousifislam) [reply surrogates/common_utils/doc_string.py]: Done!

File: surrogates/common_utils/fits.py (hunk 8 above)
Comment (tousifislam) [reply surrogates/common_utils/fits.py]: Done!

File: surrogates/common_utils/fits.py (hunk 10 above)
Comment (tousifislam) [reply surrogates/common_utils/fits.py]: Great! Changes made.

File: surrogates/common_utils/fits.py (hunk 11 above)
Comment (tousifislam) [reply surrogates/common_utils/fits.py:90]: Done!

File: surrogates/common_utils/fits.py (hunk 12 above)
Comment (tousifislam) [reply surrogates/common_utils/fits.py:151]: Yes. At this stage, for models using co-orbital frame, modes are still in the co-orbital frame.

Even evaluate _surrogate or all_modes_surrogate do not transform the mode back to inertial frame. It is done after that using utils.obtain_processed_output.
Here: 
https://github.com/tousifislam/BHPTNRSurrogate/blob/main/surrogates/model_utils/eval_surrogates.py#L93

Comment (tousifislam) [reply surrogates/common_utils/fits.py:151]: I have added the comment in the doc string as well as in the code comment

File: surrogates/model_utils/load_surrogates.py (hunk 15 above)
Comment (tousifislam) [reply surrogates/model_utils/load_surrogates.py:77]: Maybe we skip this for now :)

File: surrogates/common_utils/utils.py (hunk 13 above)
Comment (tousifislam) [reply surrogates/common_utils/utils.py]: I am fine with that

File: surrogates/common_utils/utils.py (hunk 14 above)
Comment (tousifislam) [reply surrogates/common_utils/utils.py:48]: Fine  to me

**Comment by tousifislam** [comment]
Taken care of all the comments!

**Review by sfield17 — APPROVED** [review]
